### PR TITLE
Fix watch.go behavior for watched directories

### LINKED ISSUE #531: osx: sublime test failure (test_toxt)
The issue is with "reload_test.py" where it looks (to me) like the "test_toxt" command doesn't run and the view's contents remain empty instead thus failing the assert. 

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/10352202-osx-sublime-test-failure-test_toxt?utm_campaign=plugin&utm_content=tracker%2F282001&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282001&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by jivid**
If a fsnotify event was received for a file under a directory that was being watched, a callback was only being fired for file change. In the reload_test.py, we were expecting a file creation callback to get fired, but it never was. Fixes #531

### CODE REVIEW

File: backend/watch/watch.go [hunk 1]
```diff
@@ -223,14 +223,12 @@ func (w *Watcher) Observe() {
...
 				}
-				// We will apply parent directory FileChanged callbacks to,
-				// if one of the files inside the directory has changed
-				if cbs, exist := w.watched[dir]; ev.Op&fsnotify.Write != 0 && exist {
-					for _, cb := range cbs {
-						if c, ok := cb.(FileChangedCallback); ok {
-							c.FileChanged(dir)
-						}
-					}
+				// It's possible that we're only watching the parent
```
Comment (zoli) [inline backend/watch/watch.go]: How is this comment related to the changed code?

File: backend/watch/watch.go [hunk 2]
```diff
@@ -223,14 +223,12 @@ func (w *Watcher) Observe() {
...
+				// directory of the file for which the event fired.
+				// Try to call the right callback if that's the case
+				if _, exist := w.watched[dir]; exist {
+					ev.Name = dir
+					w.apply(ev)
```
Comment (zoli) [inline backend/watch/watch.go]: I think we should just apply the `FileChangedCallback` if a file is created under some directory we should not apply `FileCreatedCallback` for the directory.

Comment (zoli) [inline backend/watch/watch.go]: Whether a file is created, deleted or renamed we should apply `FileChangedCallback` IMO.

Comment (jivid) [inline backend/watch/watch.go]: There's all these granular interfaces defined for filesystem changes, why would we not want to use them? If, however, we dostill  want a single callback for all events under a directory, I suggest the creation of a separate ```ContentsChangedCallback``` to avoid semantic conflict with ```FileChangedCallback```

Comment (zoli) [inline backend/watch/watch.go]: As far as I know for a directory there is no `fsnotify.Write` event so we can apply `FileChangedCallback` when the directory content changes(i.e when a file is created inside the directory) but applying `FileRemoved` callback on a directory when a file is removed inside the directory is inconsistent IMO.

Comment (quarnster) [inline backend/watch/watch.go]: IMO `FileCreatedCallback` should be called with the argument being the file *and* `FileChangedCallback` should be called with the argument being the dir. What do you think?

In any case we should document the callbacks to clarify when and how each of them should be used, I've opened up issue #538 to deal with that.

**Comment by jivid** [comment]
I think I messed something up while syncing changes to my fork. Working on figuring out what happened here.

**Comment by erbridge** [comment]
Please reopen this PR in https://github.com/limetext/lime-backend.
